test/tls-provider.c

```diff
@@ -748,89 +1288,2265 @@ static const OSSL_DISPATCH xor_keymgmt_functions[] = {
+    { OSSL_FUNC_KEYMGMT_GEN_INIT, (void (*)(void))xor_gen_init },
+    { OSSL_FUNC_KEYMGMT_GEN_SET_PARAMS, (void (*)(void))xor_gen_set_params },
+    { OSSL_FUNC_KEYMGMT_GEN_SETTABLE_PARAMS,
+      (void (*)(void))xor_gen_settable_params },
+    { OSSL_FUNC_KEYMGMT_GEN, (void (*)(void))xor_xorhmacsha2sig_gen },
+    { OSSL_FUNC_KEYMGMT_GEN_CLEANUP, (void (*)(void))xor_gen_cleanup },
+    { OSSL_FUNC_KEYMGMT_GET_PARAMS, (void (*) (void))xor_get_params },
+    { OSSL_FUNC_KEYMGMT_GETTABLE_PARAMS, (void (*) (void))xor_gettable_params },
+    { OSSL_FUNC_KEYMGMT_SET_PARAMS, (void (*) (void))xor_set_params },
+    { OSSL_FUNC_KEYMGMT_SETTABLE_PARAMS, (void (*) (void))xor_settable_params },
+    { OSSL_FUNC_KEYMGMT_HAS, (void (*)(void))xor_has },
+    { OSSL_FUNC_KEYMGMT_DUP, (void (*)(void))xor_dup },
+    { OSSL_FUNC_KEYMGMT_FREE, (void (*)(void))xor_freekey },
+    { OSSL_FUNC_KEYMGMT_IMPORT, (void (*)(void))xor_import },
+    { OSSL_FUNC_KEYMGMT_IMPORT_TYPES, (void (*)(void))xor_import_types },
+    { OSSL_FUNC_KEYMGMT_EXPORT, (void (*)(void))xor_export },
+    { OSSL_FUNC_KEYMGMT_EXPORT_TYPES, (void (*)(void))xor_export_types },
+    { OSSL_FUNC_KEYMGMT_LOAD, (void (*)(void))xor_load },
+    { OSSL_FUNC_KEYMGMT_MATCH, (void (*)(void))xor_match },
+    { 0, NULL }
+};
+
+typedef enum {
+    KEY_OP_PUBLIC,
+    KEY_OP_PRIVATE,
+    KEY_OP_KEYGEN
+} xor_key_op_t;
+
+/* Re-create XORKEY from encoding(s): Same end-state as after key-gen */
+static XORKEY *xor_key_op(const X509_ALGOR *palg,
+                          const unsigned char *p, int plen,
+                          xor_key_op_t op,
+                          OSSL_LIB_CTX *libctx, const char *propq)
+{
+    XORKEY *key = NULL;
+    int nid = NID_undef;
+
+    if (palg != NULL) {
+        int ptype;
+
+        /* Algorithm parameters must be absent */
+        X509_ALGOR_get0(NULL, &ptype, NULL, palg);
+        if (ptype != V_ASN1_UNDEF || !palg || !palg->algorithm) {
+            ERR_raise(ERR_LIB_USER, XORPROV_R_INVALID_ENCODING);
+            return 0;
+        }
+        nid = OBJ_obj2nid(palg->algorithm);
+    }
+
+    if (p == NULL || nid == EVP_PKEY_NONE || nid == NID_undef) {
+        ERR_raise(ERR_LIB_USER, XORPROV_R_INVALID_ENCODING);
+        return 0;
+    }
+
+    /* TBC: Check NID? */
+    key = xor_newkey(NULL);
+    if (key == NULL) {
+        ERR_raise(ERR_LIB_USER, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    if (XOR_KEY_SIZE != plen) {
+        ERR_raise(ERR_LIB_USER, XORPROV_R_INVALID_ENCODING);
+        goto err;
+    }
+
+    if (op == KEY_OP_PUBLIC) {
+        memcpy(key->pubkey, p, plen);
+	key->haspubkey = 1;
+    } else {
+        /* int i; */
+
+        memcpy(key->privkey, p, plen);
+	key->hasprivkey = 1;
+	/* XXX do not re-generate public part here to test
+	 * https://github.com/openssl/openssl/pull/19505
+        for (i = 0; i < XOR_KEY_SIZE; i++)
+            key->pubkey[i] = key->privkey[i] ^ private_constant[i];
+	key->haspubkey = 1;
+	 */
```

Comment:
...

(also, tabs → spaces)

Reply:
Actually, do you mind I completely remove this comment & code? It served the purpose -- to check #19505: With that closed, the comment and code are irrelevant, no?

Reply:
I don't mind at all if you remove that entire comment.  That will certainly fix this issue as well :wink:

Reply:
removed in https://github.com/openssl/openssl/pull/19312/commits/0330203fc25c8b2d08c052ef8070ab6f25d17209